staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go

```diff
@@ -638,6 +648,217 @@ func testHandlerConversion(t *testing.T, enableWatchCache bool) {
+				delegate: delegate,
+				validator: unstructuredSchemaCoercer{
+					dropInvalidMetadata: true,
+					repairGeneration:    true,
+					structuralSchemas:   structuralSchemas,
+					structuralSchemaGK: schema.GroupKind{
+						Group: "stable.example.com",
+						Kind:  "MultiVersion",
+					},
+					returnUnknownFieldPaths: tc.strictDecoding,
+					preserveUnknownFields:   tc.preserveUnknownFields,
+				},
+			}
+
+			obj, _, err := decoder.Decode([]byte(tc.body), nil, nil)
+			if obj != nil {
+				unstructured, ok := obj.(*unstructured.Unstructured)
+				if !ok {
+					t.Fatalf("obj is not an unstructured: %v", obj)
+				}
+				objBytes, err := unstructured.MarshalJSON()
+				if err != nil {
+					t.Fatalf("err marshaling json: %v", err)
+				}
+				expectedBytes, err := tc.expectedObj.MarshalJSON()
+				if err != nil {
+					t.Fatalf("err marshaling json: %v", err)
+				}
+				if bytes.Compare(objBytes, expectedBytes) != 0 {
+					t.Fatalf("expected obj: \n%v\n got obj: \n%v\n", tc.expectedObj, obj)
+				}
+			}
+			if err == nil || tc.expectedErr == nil {
+				if err != nil || tc.expectedErr != nil {
+					t.Fatalf("expected err: %v, got err: %v", tc.expectedErr, err)
+				}
+			} else if err.Error() != tc.expectedErr.Error() {
+				t.Fatalf("expected err: \n%v\n got err: \n%v\n", tc.expectedErr, err)
+			}
+		})
+	}
+
+}
+
+func generateStructuralSchema(crd *v1.CustomResourceDefinition, version string) (*structuralschema.Structural, error) {
+	val, err := apiextensionshelpers.GetSchemaForVersion(crd, version)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil, fmt.Errorf("the server could not properly serve the CR schema")
+	}
+	if val == nil {
+		return nil, nil
+	}
+	internalValidation := &apiextensionsinternal.CustomResourceValidation{}
+	if err := apiextensionsv1.Convert_v1_CustomResourceValidation_To_apiextensions_CustomResourceValidation(val, internalValidation, nil); err != nil {
+		return nil, fmt.Errorf("failed converting CRD validation to internal version: %v", err)
+	}
+	s, err := structuralschema.NewStructural(internalValidation.OpenAPIV3Schema)
+	if crd.Spec.PreserveUnknownFields == false && err != nil {
+		// This should never happen. If it does, it is a programming error.
+		utilruntime.HandleError(fmt.Errorf("failed to convert schema to structural: %v", err))
+		return nil, fmt.Errorf("the server could not properly serve the CR schema") // validation should avoid this
+	}
+
+	if crd.Spec.PreserveUnknownFields == false {
+		// we don't own s completely, e.g. defaults are not deep-copied. So better make a copy here.
+		s = s.DeepCopy()
+
+		if err := structuraldefaulting.PruneDefaults(s); err != nil {
+			// This should never happen. If it does, it is a programming error.
+			utilruntime.HandleError(fmt.Errorf("failed to prune defaults: %v", err))
+			return nil, fmt.Errorf("the server could not properly serve the CR schema") // validation should avoid this
+		}
+	}
+	return s, nil
+
+}
```

Comment:
It looks like this whole function is trying to undo something that you didn't have to do in the first place. I think you could extract the structural schema directly from the CRD and proceed like if the CRD had never existed?

Reply:
fixed